plutus-chain-index/src/Plutus/ChainIndex/UtxoState.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE DerivingVia           #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-| The UTXO state, kept in memory by the chain index.
+-}
+module Plutus.ChainIndex.UtxoState(
+    UtxoState
+    , UtxoIndex
+    ) where
+
+import           Data.FingerTree         (FingerTree, Measured (..))
+import           Data.Monoid             (Last (..))
+import           Data.Set                (Set)
+import qualified Data.Set                as Set
+import           GHC.Generics            (Generic)
+import           Ledger                  (Slot, TxOutRef)
+import           Plutus.ChainIndex.Types (BlockId)
+
+type UtxoIndex = FingerTree UtxoState UtxoState
+instance Measured UtxoState UtxoState where
+    measure = id
+
+-- | UTXO / ledger state, kept in memory. We are only interested in the UTXO set, everything else is stored
+--   on disk. This is OK because we don't need to validate transactions when they come in.
+data UtxoState =
+    UtxoState
+        { _usUnspentOutputs        :: Set TxOutRef -- ^ Outputs that have not been spent yet
+        , _usUnmatchedSpentOutputs :: Set TxOutRef -- ^ Outputs that we have seen a spending transaction for, but not a producing transaction
+        , _usSlot                  :: Last Slot -- ^ Last slot that we have seen
+        , _usBlockId               :: Last BlockId -- ^ Last block ID
+        }
+        deriving stock (Eq, Show, Generic)
+
+instance Semigroup UtxoState where
+    l <> r =
+        -- (<>) works like an inverse semigroup on the _usUnspentOutputs and _usUnmatchedSpentOutputs fields
+        -- it is not commutative.
+        -- TODO: Add tests
+        let newlySpent = _usUnmatchedSpentOutputs r `Set.difference` _usUnmatchedSpentOutputs l in
+        UtxoState
+            { _usUnspentOutputs = (_usUnspentOutputs l <> _usUnspentOutputs r) `Set.difference` newlySpent
+            , _usUnmatchedSpentOutputs = _usUnmatchedSpentOutputs l <> (_usUnmatchedSpentOutputs r `Set.difference` newlySpent)
```

Comment:
I drew a venn diagram and this seems to be the same as just `_usUnmatchedSpentOutputs l`. `(_usUnmatchedSpentOutputs r `Set.difference` newlySpent)` is the part of `r` that is already in `l`.

I would have thought this would actually be commutative. (`<>` as a whole isn't because of the slot and blockId fields)

```
spent = (spent l - unspent r) <> (spent r - unspent l)
unspent = (unspent l - spent r) <> (unspent r - spent l)
```

Reply:
Fine I'll add a test ;) 

I don't think it can be commutative - inputs must be produced *before* they can be spent, so every TxOutRef must eventually appear in a `_usUnspentOutputs` field before it appears in an `_usUnmatchedSpentOutputs`.

Reply:
That just means there can't be any overlap between `spent l` and `unspent r`? So we can simplify to

```
spent = spent l <> (spent r - unspent l)
unspent = (unspent l - spent r) <> unspent r
```

and then it is no longer commutative using the assumption.